gossip/src/contact_info.rs

```diff
@@ -0,0 +1,293 @@
+    #[serde(with = "short_vec")]
+    sockets: Cow<'a, [(u8, u8, u16)]>,
+}
+
+impl ContactInfo {
+    #[inline]
+    pub fn pubkey(&self) -> &Pubkey {
+        &self.pubkey
+    }
+
+    #[inline]
+    pub(crate) fn wallclock(&self) -> u64 {
+        self.wallclock
+    }
+
+    fn get_socket(&self, tag: u8) -> Option<SocketAddr> {
+        let &(_, index, port) = self.sockets.iter().find(|(k, _, _)| *k == tag)?;
+        let addr = self.addrs.get(usize::from(index))?;
+        let socket = SocketAddr::new(*addr, port);
+        let _ = sanitize_socket(&socket).ok()?;
+        Some(socket)
+    }
+
+    // Adds given IP address to self.addrs returning respective index.
+    fn push_addr(&mut self, addr: IpAddr) -> Result<u8, Error> {
+        match self.addrs.iter().position(|k| k == &addr) {
+            Some(index) => u8::try_from(index).map_err(|_| Error::IpAddrsSaturated),
+            None => {
+                let index = u8::try_from(self.addrs.len()).map_err(|_| Error::IpAddrsSaturated)?;
+                self.addrs.push(addr);
+                Ok(index)
+            }
+        }
+    }
+
+    fn set_socket(&mut self, tag: u8, socket: SocketAddr) -> Result<(), Error> {
+        let _ = sanitize_socket(&socket)?;
+        self.remove_socket(tag);
+        let addr = self.push_addr(socket.ip())?;
+        self.sockets.push((tag, addr, socket.port()));
+        if let Some(entry) = self.cache.get_mut(usize::from(tag)) {
+            *entry = Some(socket);
+        }
+        Ok(())
+    }
+
+    // Removes the socket with the specified tag.
+    fn remove_socket(&mut self, tag: u8) {
+        if let Some(index) = self.sockets.iter().position(|(k, _, _)| *k == tag) {
+            let (_, index, _) = self.sockets.remove(usize::from(index));
+            self.remove_addr(index);
+            if let Some(entry) = self.cache.get_mut(usize::from(tag)) {
+                *entry = None;
+            }
+        }
+    }
+
+    // If no socket uses the IP address at the given index,
+    // removes the IP address and adjusts indices.
+    fn remove_addr(&mut self, index: u8) {
+        if !self.sockets.iter().any(|(_, k, _)| *k == index) {
+            self.addrs.remove(usize::from(index));
+            for (_, k, _) in &mut self.sockets {
+                if *k > index {
+                    *k -= 1;
+                }
+            }
+        }
+    }
+}
+
+fn sanitize_socket(socket: &SocketAddr) -> Result<(), Error> {
+    if socket.port() == 0u16 {
+        return Err(Error::InvalidPort(socket.port()));
+    }
+    let addr = socket.ip();
+    if addr.is_unspecified() {
+        return Err(Error::UnspecifiedAddr(addr));
+    }
+    if addr.is_multicast() {
+        return Err(Error::MulticastAddr(addr));
+    }
+    Ok(())
+}
+
+impl Sanitize for ContactInfo {
+    fn sanitize(&self) -> Result<(), SanitizeError> {
+        if self.wallclock >= MAX_WALLCLOCK {
+            return Err(SanitizeError::ValueOutOfBounds);
+        }
+        Ok(())
```

Comment:
we should validate the address indexes, socket tags and ports here too

Reply:
Those are validated earlier during deserialization;
```
deserialize => ContactInfo::try_from => sanitize_entries
```